File: clase30sep/farmacia.py
from multipledispatch import dispatch

class Farmacia:
    def __init__(self, nombre="", cant=0, matMed=[], trab = []):
        self.__nombre = nombre
        self.__nMedicamnetos = cant
        self.__medis = matMed
        self.__trabajadores = trab
    
    def getMedis(self):
        return self.__medis
    
    def getNMedicamentos(self):
        return self.__nMedicamnetos
    
    def __add__(self, otro):
        matrizMedisOtro = otro.getMedis()
        nombreX, tipoX = matrizMedisOtro[0][0], matrizMedisOtro[0][1]
        
        for i in range(0, self.__nMedicamnetos, 1):
            if(nombreX == self.__medis[i][0] and tipoX == self.__medis[i][1]):
                return True
        return False
    
    def __sub__(self, otro):
        cantidadIncremento = otro.getNMedicamentos()
        
        for medicamento in self.__medis:
            if(medicamento[2] < 10):
                medicamento[2] = medicamento[2]+cantidadIncremento
                
        return 0
    
    def setMedis(self, medis):
        self.__medis = medis
    
    def getNombre(self):
        return self.__nombre

    def mostrar(self):
        print(f"-.-.-.-.-.-{self.__nombre}-.-.-.-.-.-")
        print(f"Cantidad de Medicamento: {self.__nMedicamnetos}")
        print("---Medicamentos---")
        for medicamento in self.__medis:
            print("", end= "\t")
            print(medicamento)
            
        for trabajador in self.__trabajadores:
            print("", end= "\t")
            print(trabajador)
    
    @dispatch()
    def metodoSobrecargado(self):
        copia = self.getMedis()
        ordenado = []
        
        while(len(copia)!=0):
            masPequeño = []
            asciiPeque = 96
            porVer = []
            for medi in copia:
                if(ord(medi[0][0]) > asciiPeque):
                    masPequeño = medi
                    asciiPeque = ord(medi[0][0])
                    
                    
            for medi in copia:
                if(masPequeño != medi):
                    porVer.append(medi)
                else:
                    ordenado.append(masPequeño)
            copia = porVer
            
        self.setMedis(ordenado)
    
    @dispatch(str)
    def metodoSobrecargado(self, nombreX):
        for trabajador in self.__trabajadores:
            if(nombreX == trabajador[0]):
                print(f" El trabajador {nombreX} esta en el turno de {trabajador[1]}")
                return 0
        
        print(f"No se a encontrado el trabajor {nombreX}")
        return 0

    @dispatch(object, str)
    def metodoSobrecargado(self, otro, tipoX):
        cantidad1 = 0
        cantidad2 = 0
        
        for medicamento in self.__medis:
            if(medicamento[1] == tipoX):
                cantidad1 += medicamento[2]
        
        for medicamento in otro.getMedis():
            if(medicamento[1] == tipoX):
                cantidad2 += medicamento[2]
        
        if(cantidad1 == cantidad2):
            print("AMBOS TIENEN LA MISMA CANTIDAD DE MEDICAMENTOS DE TIPO ", tipoX)
            print(cantidad1, cantidad2)
        elif(cantidad1 > cantidad2):
            print(self.__nombre)
        else:
            print(otro.getNombre())
        

f1 = Farmacia(nombre = "Farmcia Bolivia", trab = [["Raul", "noche"], ["Jose", "mañana"], ["Pepe", "tarde"]])
medis = [
    ["ibuprofeno", "pastilla", 4],
    ["paracetamol", "pastilla", 1],
    ["diclofenaco", "pomada", 14]
]

medis1 = [
    ["ibuprofeno", "pastilla", 4],
    ["paracetamol", "pastilla", 10],
    ["diclofenaco", "pomada", 14]
]

f2 = Farmacia(nombre="Farmacias Virgen de Copacabana", cant= 3, matMed= medis1)

f3 = Farmacia(nombre = "farmacia Chavez", cant = 3, matMed= medis, trab = [["Raul", "noche"], ["Jose", "mañana"], ["Pepe", "tarde"]])
f4 = Farmacia(matMed= [["diclofenaco", "pastilla"]])

seEncontro = f3+f4

print(seEncontro)

f3.mostrar()

print("\n\n")
f5 = Farmacia(cant=100)
# f3 - f5

f3.mostrar()

print("------------")
f3.metodoSobrecargado()
f3.metodoSobrecargado("POncio")
f3.metodoSobrecargado(f2, "pastilla")

File: clase30sep/clases/Farmacia.java
package clases;


public class Farmacia {

    private String nombre;
    private int nMedicamentos;
    private String medis[][];
    private int disponible[];
    private String trabajadores[][];

    public Farmacia(String nombre, int cant, String matmed[][], int dispo[], String trab[][]) {
        this.nombre = nombre;
        this.nMedicamentos = cant;
        this.medis = matmed;
        this.disponible = dispo;
        this.trabajadores = trab;
    }

    public Farmacia(String nombre, int cant, String trab[][]) {
        this.nombre = nombre;
        this.nMedicamentos = cant;
        this.medis = new String[10][2];
        this.disponible = new int[10];
        this.trabajadores = trab;
    }

    public Farmacia(String nombre, int cant, String matmed[][], int dispo[]) {
        this.nombre = nombre;
        this.nMedicamentos = cant;
        this.medis = matmed;
        this.disponible = dispo;
        this.trabajadores = new String[10][2];
    }

    public Farmacia(String medi[][]) {
        this.nombre = "";
        this.nMedicamentos = 0;
        this.medis = medi;
        this.disponible = new int[1];
        this.trabajadores = new String[2][2];
    }

    public Farmacia(int cant) {
        this.nombre = "";
        this.nMedicamentos = cant;
        this.medis = new String[1][1];
        this.disponible = new int[1];
        this.trabajadores = new String[2][2];
    }

    public String[][] getMedis() {
        return this.medis;
    }

    public int getNMedicamentos() {
        return this.nMedicamentos;
    }

    public boolean suma(Farmacia otro) {
        String mediBuscado[][] = otro.getMedis();
        String nombreX = mediBuscado[0][0];
        String tipoX = mediBuscado[0][1];

        for (int i = 0; i < this.nMedicamentos; i++) {
            if (nombreX.equals(this.medis[i][0]) && tipoX.equals(this.medis[i][1])) {
                return true;
            }
        }
        return false;
    }

    public int resta(Farmacia otro) {
        int incremento = otro.getNMedicamentos();

        for (int i = 0; i < this.nMedicamentos; i++) {
            if (this.disponible[i] < 10) {
                this.disponible[i] = this.disponible[i] + incremento;
            }
        }
        return 0;
    }

    public void mostrar() {
        System.out.println("-.-.-.-.-.-" + this.nombre);
        System.out.println("Cantidad de Medicamentos: " + this.nMedicamentos);
        for (int i = 0; i < this.nMedicamentos; i++) {
            System.out.println(this.medis[i][0] + " " + this.medis[i][1] + " " + this.disponible[i]);
        }

        System.out.println(".-.-.-.-TRABAJADORES-.-.-.-.");
        for (int i = 0; i < this.nMedicamentos; i++) {
            System.out.println(this.trabajadores[i][0] + " " + this.trabajadores[i][1]);
        }
    }

    public int[] getDisponible() {
        return this.disponible;
    }

    public void setMedis(String medis[][]) {
        this.medis = medis;
    }

    public void setDisponible(int dispo[]) {
        this.disponible = dispo;
    }

    public String getNombre(){
        return  this.nombre;
    }

    public void metodoSobrecargado() {
        String copia[][] = this.getMedis();
        int copiaDispo[] = this.getDisponible();
        String ordenado[][] = new String[this.nMedicamentos][2];
        int ordDipo[] = new int[this.nMedicamentos];
        int j = 0;
        int n = this.nMedicamentos;

        while (copia.length != 0) {
            n--;
            String masPequeño[] = {};
            int dispoMed = 0;
            int ascciPeque = 96;
            String porVer[][] = new String[n][2];
            int porVerDispo[] = new int[n];
            for (int k = 0; k < copia.length; k++) {
                if ((int) copia[k][0].charAt(0) > ascciPeque) {
                    masPequeño = new String[]{copia[k][0], copia[k][1]};
                    dispoMed = copiaDispo[k];
                    ascciPeque = (int) copia[k][0].charAt(0);
                }
            }
            int f = 0;
            for (int k = 0; k < copia.length; k++) {
                if (copia[k][0].equals(masPequeño[0])) {
                    ordenado[j] = masPequeño;
                    ordDipo[j] = dispoMed;
                    j++;
                } else {
                    porVer[f] = copia[k];
                    porVerDispo[f] = copiaDispo[k];
                    f++;
                }
            }
            copia = porVer;
            copiaDispo = porVerDispo;
        }

        this.setDisponible(ordDipo);
        this.setMedis(ordenado);
    }

    public void metodoSobrecargado(String nombreX){
        boolean seEncontro = false;
        for(int i = 0; i < this.trabajadores.length; i++){
            if(nombreX.equals(this.trabajadores[i][0])){
                System.out.println(" El trabajador "+nombreX+" esta en el turno de " + this.trabajadores[i][1]);
                seEncontro = true;
                break;
            }
        }
        if(! seEncontro){
            System.out.println("No se a encontrado el trabajor "+nombreX);
        }
    }

    public void metodoSobrecargado(Farmacia otro, String tipoX){
        int cantidad1 = 0;
        int cantidad2 = 0;
        for (int i = 0; i < this.nMedicamentos; i++){
            if(this.medis[i][1].equals(tipoX)){
                cantidad1 += this.disponible[i];
            }
        }

        for (int i = 0; i < otro.getNMedicamentos(); i++){
            if(otro.getMedis()[i][1].equals(tipoX)){
                cantidad2 += otro.getDisponible()[i];
            }
        }

        if(cantidad1 == cantidad2){
            System.out.println("AMBAS FARMACIAS TIENEN LA MISMCA CANTIDAD DE MEDICAMENTOS DE TIPO "+ tipoX);
        }else if(cantidad1 > cantidad2){
            System.out.println(this.nombre);
        }else{
            System.out.println(otro.getNombre());
        }
    }

    public static void main(String[] args) {

        String medis[][] = {
            {"ibuprofreno", "pastillas"},
            {"aracetamol", "pastillas"},
            {"diclofenaco", "pomada"}
        };
        int dispo[] = {
            6, 25, 4
        };

        String medis1[][] = {
            {"ibuprofreno", "pastillas"},
            {"paracetamol", "pastillas"},
            {"diclofenaco", "pomada"}
        };
        int dispo1[] = {
            66, 25, 4
        };

        Farmacia f1 = new Farmacia("farmacias Bolivia", 3, medis1, dispo1);
        String farmas[][] = {
            {"Jorge", "mañana"},
            {"Jose", "noche"},
            {"Raul", "tarde"}
        };
        Farmacia f2 = new Farmacia("Farmacia Virgen de Coapacabana", 0, farmas);

        Farmacia f3 = new Farmacia("Farmacia Chavez", 3, medis, dispo, farmas);
        String med[][] = {{"ibuprofreno", "jarabe"}};
        Farmacia f4 = new Farmacia(med);

        boolean seEncontro = f3.suma(f4);
        System.out.println(seEncontro);

        f3.mostrar();

        Farmacia f5 = new Farmacia(14);
        //f3.resta(f5);

        System.out.println("\n\nDESPUES\n\n");

        System.out.println("...................");
        f3.metodoSobrecargado();
        f3.mostrar();
        f3.metodoSobrecargado("POncio");
        f3.metodoSobrecargado(f1, "pastillas");
    }

}
